### ISSUE TITLE
Inconsistency in tf.nn.top_k when using sorted=False and tf.float32 tensor and GPU placement

### ISSUE THREAD

**Issue by jhultman** [tensorflow/tensorflow#22383]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No.
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04.
- **Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device**: NA.
- **TensorFlow installed from (source or binary)**: Binary.
- **TensorFlow version (use command below)**: 'v1.10.1-0-g4dcfddc5d1', '1.10.1'
- **Python version**: Python 3.6.5 :: Anaconda, Inc.
- **Bazel version (if compiling from source)**: NA
- **GCC/Compiler version (if compiling from source)**: NA
- **CUDA/cuDNN version**: CUDAToolkit 9.0, CuDNN 7.1.4
- **GPU model and memory**: NVIDIA Quadro P4000, 8GB (driver version 396.37)
- **Exact command to reproduce**: See source code below.

### Describe the problem
The indices and values returned by top_k(sorted=False) seem to be shuffled with respect to each other when tf.float32 tensors are used, but not when tf.int32 tensors are used. I understand that we can not expect the values to be returned in ascending/descending order since we've set sorted=False, but shouldn't the values and indices at least be consistently ordered with respect to each other? Hope this makes sense. See my minimum working example below.

Edit: This seems to be a GPU-only problem.

### Source code / logs

```
import tensorflow as tf
import numpy as np
tf.set_random_seed(21)

def demo_bug(np_dtype, tf_dtype):
    np_arr = np.array([3, 2, 8, 1], np_dtype)
    arr = tf.convert_to_tensor(np_arr, tf_dtype)
    top_k = tf.nn.top_k(arr, k=3, sorted=False)

    with tf.Session() as sess:
        vals, inds = sess.run(top_k)

    print('dtype:   ', tf_dtype.__repr__())
    print('arr:     ', np_arr)
    print('inds:    ', inds)
    print('tf_vals: ', vals)
    print('np_vals: ', np_arr[inds])

demo_bug(np.int32, tf.int32)
demo_bug(np.float32, tf.float32)

>> dtype:    tf.int32
>> arr:      [3 2 8 1]
>> inds:     [2 0 1]
>> tf_vals:  [8 3 2]
>> np_vals:  [8 3 2]

>> dtype:    tf.float32
>> arr:      [3. 2. 8. 1.]
>> inds:     [2 0 1]
>> tf_vals:  [2. 3. 8.]
>> np_vals:  [8. 3. 2.]
```

**Comment by harshini-gadige** [comment]
@jhultman Hi, I've executed your code and I get the values and indices in a consistent order. Please find the result below.

('dtype:   ', 'tf.int32')
('arr:     ', array([3, 2, 8, 1], dtype=int32))
('inds:    ', array([1, 0, 2], dtype=int32))
('tf_vals: ', array([2, 3, 8], dtype=int32))
('np_vals: ', array([2, 3, 8], dtype=int32))
('dtype:   ', 'tf.float32')
('arr:     ', array([3., 2., 8., 1.], dtype=float32))
('inds:    ', array([1, 0, 2], dtype=int32))
('tf_vals: ', array([2., 3., 8.], dtype=float32))
('np_vals: ', array([2., 3., 8.], dtype=float32))

**Comment by jhultman** [comment]
I just tried it again in a different virtualenv installed from source (not pip package) with tf.__version__ == 1.9.0 and I was able to reproduce the bug. My colleague was not able to reproduce the bug despite being on 16.04 with tf v1.10.1 latest pip package.

I tried again with a different test case and same problem:

```
dtype:    tf.float32
arr:      [3. 2. 8. 6.]
inds:     [2 3 0]
tf_vals:  [3. 6. 8.]
np_vals:  [8. 6. 3.]
```

**Comment by harshini-gadige** [comment]
@jhultman Okay. The issue is with tensorflow v 1.9. I recommend to use tensorflow v 1.10 for this op.

**Comment by jhultman** [comment]
Sorry, I think maybe you misunderstood. I am able to reproduce the bug in _both_ v 1.9.0 and v 1.10.1.

**Comment by jhultman** [comment]
I tried with different device placements and it seems like a GPU-only bug for tf.float32 tensors. Maybe someone else can try with GPU placement and see if they can reproduce the bug. See my updated dummy example below.

`import tensorflow as tf
import numpy as np
import itertools
tf.set_random_seed(21)

def demo_bug(np_dtype, tf_dtype, device):
    with tf.device(device):
        np_arr = np.array([3, 2, 8, 1], np_dtype)
        arr = tf.convert_to_tensor(np_arr, tf_dtype)
        top_vals, top_inds = tf.nn.top_k(arr, k=3, sorted=False)

    with tf.Session() as sess:
        vals, inds = sess.run((top_vals, top_inds))

    print('device:     ', device)
    print('dtype:      ', tf_dtype.__repr__())
    print('arr:        ', np_arr)
    print('inds:       ', inds)
    print('tf_vals:    ', vals)
    print('np_vals:    ', np_arr[inds], end='\n\n')

int_dtypes = (np.int32, tf.int32)
float_dtypes = (np.float32, tf.float32)

demo_bug(*int_dtypes, '/cpu')
demo_bug(*float_dtypes, '/cpu')
demo_bug(*int_dtypes, '/gpu')
demo_bug(*float_dtypes, '/gpu')

>> device:      /cpu
>> dtype:       tf.int32
>> arr:         [3 2 8 1]
>> inds:        [1 0 2]
>> tf_vals:     [2 3 8]
>> np_vals:     [2 3 8]
 
>> device:      /cpu
>> dtype:       tf.float32
>> arr:         [3. 2. 8. 1.]
>> inds:        [1 0 2]
>> tf_vals:     [2. 3. 8.]
>> np_vals:     [2. 3. 8.]

>> device:      /gpu
>> dtype:       tf.int32
>> arr:         [3 2 8 1]
>> inds:        [2 0 1]
>> tf_vals:     [8 3 2]
>> np_vals:     [8 3 2]

>> device:      /gpu
>> dtype:       tf.float32
>> arr:         [3. 2. 8. 1.]
>> inds:        [2 0 1]
>> tf_vals:     [2. 3. 8.]
>> np_vals:     [8. 3. 2.]
`

**Comment by jhultman** [comment]
Anybody interested in this?

**Comment by martinwicke** [comment]
This looks like a bug in the gpu implementation of top_k. @chsigg could you take a look at this?

**Comment by tensorflowbutler** [comment]
Nagging Assignees @martinwicke, @chsigg: It has been 29 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by qqfish** [comment]
Thanks for reporting this. It is caused by constant folding. TF folds partial output of TopK on GPU with dtype float.

I am sending out a fix to change constant folding to either fold all the outputs of an op or not fold any outputs at all.
